spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestFunctionCatalog.java

```diff
@@ -0,0 +1,129 @@
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestFunctionCatalog extends SparkCatalogTestBase {
+
+  @Parameterized.Parameters(name = "catalogConfig = {0}")
+  public static Object[][] parameters() {
+    return new Object[][] {{SparkCatalogConfig.HADOOP}};
+  }
+
+  private static final String[] EMPTY_NAMESPACE = new String[] {};
+  private static final String[] SYSTEM_NAMESPACE = new String[] {"system"};
+  private final FunctionCatalog asFunctionCatalog;
+
+  public TestFunctionCatalog(SparkCatalogConfig catalogConfig) {
+    super(catalogConfig);
+    this.asFunctionCatalog = castToFunctionCatalog(catalogName);
+  }
+
+  @Test
+  public void testListFunctionsViaCatalog() throws NoSuchNamespaceException {
+    Assertions.assertThat(asFunctionCatalog.listFunctions(EMPTY_NAMESPACE))
+        .anyMatch(func -> "iceberg_version".equalsIgnoreCase(func.name()));
+
+    Assertions.assertThat(asFunctionCatalog.listFunctions(SYSTEM_NAMESPACE))
+        .anyMatch(func -> "iceberg_version".equalsIgnoreCase(func.name()));
+  }
+
+  @Test
+  public void testLoadAndDescribeUnboundFunctionsViaCatalog() throws NoSuchFunctionException {
+    for (String[] namespace : ImmutableList.of(EMPTY_NAMESPACE, SYSTEM_NAMESPACE)) {
+      Identifier identifier = Identifier.of(namespace, "iceberg_version");
+      UnboundFunction func = asFunctionCatalog.loadFunction(identifier);
+
+      Assertions.assertThat(func).isExactlyInstanceOf(IcebergVersionFunction.class);
+      Assertions.assertThat(func.name()).asString().isEqualTo("iceberg_version");
+      Assertions.assertThat(func.description())
+          .containsIgnoringCase("Returns the runtime Iceberg version");
+    }
+  }
+
+  @Test
+  public void testBindAndUseFunctionsViaCatalog() throws NoSuchFunctionException {
+    for (String[] namespace : ImmutableList.of(EMPTY_NAMESPACE, SYSTEM_NAMESPACE)) {
+      Identifier identifier = Identifier.of(namespace, "iceberg_version");
+      UnboundFunction unboundFunc = asFunctionCatalog.loadFunction(identifier);
+
+      BoundFunction boundFunc = unboundFunc.bind(new StructType());
+      Assertions.assertThat(boundFunc.getClass().toString())
+          .containsIgnoringCase("IcebergVersionFunctionImpl");
+      Assertions.assertThat(boundFunc.canonicalName()).isEqualTo("iceberg.iceberg_version");
+      Assertions.assertThat(boundFunc.isResultNullable()).isFalse();
+      Assertions.assertThat(boundFunc.isDeterministic()).isTrue();
+
+      // Ensure the bound result produces the expected UTF8String type.
+      ScalarFunction<UTF8String> scalarFunc = (ScalarFunction<UTF8String>) boundFunc;
+      Assert.assertEquals(
+          UTF8String.fromString(IcebergBuild.version()),
+          scalarFunc.produceResult(InternalRow.empty()));
+    }
+  }
+
+  @Test
+  public void testCannotLoadFunctionsFromInvalidNamespace() {
```

Comment:
I think this test should use `loadFunction` directly, like the others. I don't see why this one uses SQL while the other load and list tests use the API.

Reply:
Added a test that uses the catalog directly so it can also assert on the error message.